GTScript-Parallel-model.md

```diff
@@ -103,3 +103,107 @@ for k in range(start, end):
         u[i, j, k] = 1
     parfor [i_start:i_end, j_start:j_end]:
         b[i, j, k] = u[i-2,j,k] + u[i+1,j,k] + u[i,j-1,k] + u[i,j-2,k]
+```
+
+## Conditionals
+
+### If-Statements
+
+- Conditions in `if` statements are required to be of scalar type, e.g. scalar variables (not fields) or scalar expressions (no fields involved).
```

Comment:
This does not work for us.
There are very significant cases where we absolutely need this. To name a few:
- limiters 
- cfl conditions
- parametrizations that work on clouds / ice etc.

If this is implemented we can no longer work against that model

Reply:
I second this. There are three types of if-statements that I think we should support: compile-time scalars, runtime scalars, runtime fields. The use-cases for fields are numerous and @twicki mentioned some important ones. For fields, the condition should simply be that there is a i,j-dimension present, but they e.g. can be 2d or 3d.

Reply:
> This does not work for us.

No, this is not true. You still can use the ternary expression.

Reply:
And the exercise here is: find all the cases for limiters, cfl, parameterizations that have an if on field and count how many fields are written inside of these ifs. If the number is large (maybe > 2) this proposal is probably not good. If the number is 1, the proposal is probably the solution. For everything in between 1 and large we need to think about trade-offs.

Reply:
I checked every `if` in the COSMO dycore (C++), all `if`s on fields, except one, are of the form
```cpp
            if (flux >= (Real)0.0) {
                flux *= math::min((Real)1.0, eval(rxp(Offset() + 1)), eval(rxm()));
            } else {
                flux *= math::min((Real)1.0, eval(rxp()), eval(rxm(Offset() + 1)));
```

The exception is the following from saturation adjustment:
```cpp
            if ((compute_gqd(gew, phe) - qdwe) < 0) {
                gew = compute_gew(tg0);
                Real zh = CPD * eval(t_in()) + LHV * eval(qv());
                Real gqd0 = compute_gqd(gew, phe);

                for (int i = 0; i < eval(niter()) - 1; ++i) {
                    Real dqdt0 = compute_dqdt(tg0, gqd0);
                    tg0 = (zh - LHV * (gqd0 - dqdt0 * tg0)) / (CPD + LHV * dqdt0);
                    gew = compute_gew(tg0);
                    gqd0 = compute_gqd(gew, phe);
                }

                Real dqdt0 = compute_dqdt(tg0, gqd0);
                Real tgn = (zh - LHV * (gqd0 - dqdt0 * tg0)) / (CPD + LHV * dqdt0);
                gqd0 = gqd0 + dqdt0 * (tgn - tg0);

                eval(t_in()) = tgn;
                eval(qv()) = gqd0;
                eval(qc()) = gridtools::math::max(qdwe - gqd0, (Real)1e-20);
```

Reply:
Could you do the same exercise for the FV3 dycore. Probably we should also check how it looks in physics.

Reply:
And one more comment about the above and why I like this proposal (currently).

My expectation was, `if`s on fields are typically of the ternary form as in the flux example above, i.e. small `if`-bodies, while `if`s on scalars have typically huge bodies, because they are flags, changing an algorithm. The former is confirmed by looking at COSMO, I think also the latter is confirmed, but it's not fully obvious as it involves the question of which `if`s belong to the driver code outside of GTScript and which ones would be inside.

Note that going with a proposal where the parfor is outside of the if, we get in troubles as well, see the following example.
```python
with computation(...) and stuff:
    if runtime_flag:
        # a lot of code here
        tmp = a[-1,0,0]
        # a lot of code here
        a = tmp[-1,0,0]
```
This is Eddies famous example, which currently fails but (without the if) should actually be legal. Also in this proposal, this example is legal. In any proposal that puts the parfor outside, the above example has to be split
```python
with computation(...) and stuff:
    if runtime_flag:
        # a lot of code here
        tmp = a[-1,0,0]
        # a lot of code here
    if runtime_flag:
        a = tmp[-1,0,0]
```
Now add an else or some nested `if`s and you get problems.